File: src/providers/git/github.py
"""GitHub git provider adapter."""

from base64 import b64decode, b64encode
import httpx
from src.providers.base import GitProviderBase


class GitHubAdapter(GitProviderBase):
    name = "github"
    pr_label = "pull request"

    def parse_webhook(self, headers, payload, config):
        event = headers.get("x-github-event")
        bot_users = config.get("bot_users", [])

        if event == "pull_request_review":
            if payload.get("review", {}).get("state") != "approved":
                return None
            return {
                "event": "approved",
                "branch": payload.get("pull_request", {}).get("head", {}).get("ref"),
                "pr_id": payload.get("pull_request", {}).get("number"),
                "author": payload.get("review", {}).get("user", {}).get("login", ""),
            }

        if event == "pull_request":
            pr = payload.get("pull_request", {})
            if payload.get("action") == "closed" and pr.get("merged"):
                return {
                    "event": "approved",
                    "branch": pr.get("head", {}).get("ref"),
                    "pr_id": pr.get("number"),
                    "author": payload.get("sender", {}).get("login", ""),
                }
            return None

        if event == "push":
            author = payload.get("sender", {}).get("login", "")
            if author in bot_users:
                return None
            return {
                "event": "push",
                "branch": payload.get("ref", "").replace("refs/heads/", ""),
                "pr_id": None,
                "author": author,
            }

        if event in ("issue_comment", "pull_request_review_comment"):
            if event == "issue_comment" and not payload.get("issue", {}).get("pull_request"):
                return None
            author = payload.get("comment", {}).get("user", {}).get("login", "")
            if author in bot_users:
                return None
            return {
                "event": "comment",
                "branch": None,
                "pr_id": payload.get("issue", {}).get("number") or payload.get("pull_request", {}).get("number"),
                "author": author,
            }

        return None

    def create_api(self, env):
        owner = env["GITHUB_OWNER"]
        repo = env["GITHUB_REPO"]
        client = httpx.Client(
            base_url="https://api.github.com",
            headers={"Authorization": f"Bearer {env['GITHUB_TOKEN']}", "Accept": "application/vnd.github.v3+json"},
        )
        r = f"/repos/{owner}/{repo}"

        class Api:
            def create_branch(self, name, ref="main"):
                ref_data = client.get(f"{r}/git/ref/heads/{ref}").json()
                sha = ref_data["object"]["sha"]
                return client.post(f"{r}/git/refs", json={"ref": f"refs/heads/{name}", "sha": sha}).json()

            def commit_files(self, branch, message, actions):
                for action in actions:
                    if action["action"] == "delete":
                        existing = client.get(f"{r}/contents/{action['file_path']}", params={"ref": branch}).json()
                        client.request("DELETE", f"{r}/contents/{action['file_path']}", json={"message": message, "branch": branch, "sha": existing["sha"]})
                    else:
                        sha = None
                        try:
                            existing = client.get(f"{r}/contents/{action['file_path']}", params={"ref": branch}).json()
                            sha = existing.get("sha")
                        except Exception:
                            pass
                        body = {"message": message, "branch": branch, "content": b64encode((action.get("content") or "").encode()).decode()}
                        if sha:
                            body["sha"] = sha
                        client.put(f"{r}/contents/{action['file_path']}", json=body)
                return {"message": message}

            def create_pr(self, source, target, title, description):
                data = client.post(f"{r}/pulls", json={"head": source, "base": target, "title": title, "body": description}).json()
                return {"id": data.get("number"), "url": data.get("html_url"), "title": data.get("title"), "state": data.get("state")}

            def get_pr(self, pr_id):
                data = client.get(f"{r}/pulls/{pr_id}").json()
                return {"id": data.get("number"), "url": data.get("html_url"), "title": data.get("title"), "state": data.get("state"), "source_branch": data.get("head", {}).get("ref"), "description": data.get("body")}

            def update_pr(self, pr_id, updates):
                body = {}
                if "title" in updates:
                    body["title"] = updates["title"]
                if "description" in updates:
                    body["body"] = updates["description"]
                return client.patch(f"{r}/pulls/{pr_id}", json=body).json()

            def list_pr_comments(self, pr_id):
                issue = client.get(f"{r}/issues/{pr_id}/comments").json()
                review = client.get(f"{r}/pulls/{pr_id}/comments").json()
                all_comments = [
                    *[{"id": c["id"], "author": c.get("user", {}).get("login"), "body": c.get("body"), "created_at": c.get("created_at"), "system": False} for c in issue],
                    *[{"id": c["id"], "author": c.get("user", {}).get("login"), "body": c.get("body"), "created_at": c.get("created_at"), "system": False} for c in review],
                ]
                return sorted(all_comments, key=lambda x: x.get("created_at", ""))

            def post_pr_comment(self, pr_id, body):
                return client.post(f"{r}/issues/{pr_id}/comments", json={"body": body}).json()

            def get_file(self, file_path, ref="main"):
                data = client.get(f"{r}/contents/{file_path}", params={"ref": ref}).json()
                data["content"] = b64decode(data.get("content", "")).decode("utf-8")
                return data

        api = Api()
        self.validate_api(api)
        return api


adapter = GitHubAdapter()

File: src/providers/git/github.js
/**
 * @module providers/git/github
 * @description GitHub git provider adapter.
 *
 * Parses GitHub webhook events (pull_request, push, issue_comment, pull_request_review)
 * and provides an API client wrapping GitHub REST API.
 */

const axios = require('axios');
const { GitProviderBase } = require('../base/git-provider-base');

class GitHubAdapter extends GitProviderBase {
  get name() { return 'github'; }
  get prLabel() { return 'pull request'; }

  parseWebhook(headers, payload, config) {
    const event = headers['x-github-event'];
    const botUsers = config.botUsers || [];

    if (event === 'pull_request_review') {
      if (payload.review?.state !== 'approved') return null;
      return {
        event: 'approved',
        branch: payload.pull_request?.head?.ref,
        prId: payload.pull_request?.number,
        author: payload.review?.user?.login || '',
      };
    }

    if (event === 'pull_request') {
      if (payload.action === 'closed' && payload.pull_request?.merged) {
        return {
          event: 'approved',
          branch: payload.pull_request?.head?.ref,
          prId: payload.pull_request?.number,
          author: payload.sender?.login || '',
        };
      }
      return null;
    }

    if (event === 'push') {
      const author = payload.sender?.login || '';
      if (botUsers.includes(author)) return null;
      return {
        event: 'push',
        branch: (payload.ref || '').replace('refs/heads/', ''),
        prId: null,
        author,
      };
    }

    if (event === 'issue_comment' || event === 'pull_request_review_comment') {
      if (!payload.issue?.pull_request && event === 'issue_comment') return null;
      const author = payload.comment?.user?.login || '';
      if (botUsers.includes(author)) return null;
      return {
        event: 'comment',
        branch: null,
        prId: payload.issue?.number || payload.pull_request?.number,
        author,
      };
    }

    return null;
  }

  createApi(env) {
    const owner = env.GITHUB_OWNER;
    const repo = env.GITHUB_REPO;
    const client = axios.create({
      baseURL: 'https://api.github.com',
      headers: {
        Authorization: `Bearer ${env.GITHUB_TOKEN}`,
        Accept: 'application/vnd.github.v3+json',
        'Content-Type': 'application/json',
      },
    });
    const r = `/repos/${owner}/${repo}`;

    const api = {
      async createBranch(branchName, ref = 'main') {
        const { data: refData } = await client.get(`${r}/git/ref/heads/${ref}`);
        const { data } = await client.post(`${r}/git/refs`, { ref: `refs/heads/${branchName}`, sha: refData.object.sha });
        return data;
      },
      async commitFiles(branch, message, actions) {
        for (const action of actions) {
          if (action.action === 'delete') {
            const { data: existing } = await client.get(`${r}/contents/${action.file_path}?ref=${branch}`);
            await client.delete(`${r}/contents/${action.file_path}`, { data: { message, branch, sha: existing.sha } });
          } else {
            let sha;
            try { const { data: e } = await client.get(`${r}/contents/${action.file_path}?ref=${branch}`); sha = e.sha; } catch (_) {}
            const body = { message, branch, content: Buffer.from(action.content || '').toString('base64') };
            if (sha) body.sha = sha;
            await client.put(`${r}/contents/${action.file_path}`, body);
          }
        }
        return { message };
      },
      async createPR(sourceBranch, targetBranch, title, description) {
        const { data } = await client.post(`${r}/pulls`, { head: sourceBranch, base: targetBranch, title, body: description });
        return { id: data.number, url: data.html_url, title: data.title, state: data.state };
      },
      async getPR(prId) {
        const { data } = await client.get(`${r}/pulls/${prId}`);
        return { id: data.number, url: data.html_url, title: data.title, state: data.state, sourceBranch: data.head?.ref, description: data.body };
      },
      async updatePR(prId, updates) {
        const body = {};
        if (updates.title) body.title = updates.title;
        if (updates.description) body.body = updates.description;
        const { data } = await client.patch(`${r}/pulls/${prId}`, body);
        return data;
      },
      async listPRComments(prId) {
        const [{ data: issue }, { data: review }] = await Promise.all([
          client.get(`${r}/issues/${prId}/comments`),
          client.get(`${r}/pulls/${prId}/comments`),
        ]);
        return [
          ...issue.map(c => ({ id: c.id, author: c.user?.login, body: c.body, createdAt: c.created_at, system: false })),
          ...review.map(c => ({ id: c.id, author: c.user?.login, body: c.body, createdAt: c.created_at, system: false })),
        ].sort((a, b) => new Date(a.createdAt) - new Date(b.createdAt));
      },
      async postPRComment(prId, body) {
        const { data } = await client.post(`${r}/issues/${prId}/comments`, { body });
        return data;
      },
      async getFile(filePath, ref = 'main') {
        const { data } = await client.get(`${r}/contents/${filePath}`, { params: { ref } });
        return { ...data, content: Buffer.from(data.content, 'base64').toString('utf-8') };
      },
    };

    this.validateApi(api);
    return api;
  }
}

module.exports = new GitHubAdapter();
